### PR TITLE
Remove project type from Language class

### PR DESCRIPTION
**Pull request by duncanp-sonar**
* language -> project is not a 1-1 mapping

### CODE REVIEW

File: src/Integration.UnitTests/LanguageTests.cs [hunk 1]
```diff
@@ -47,77 +46,30 @@ public void Language_Ctor_GuidString_ArgChecks()
...
             // Act + Assert
-            foreach(var supportedLang in Language.SupportedLanguages)
+            foreach (var supportedLang in Language.SupportedLanguages)
             {
                 supportedLang.IsSupported.Should().BeTrue("Supported language should be supported");
             }
         }
 
         [TestMethod]
-        public void Language_ForProject_KnownLanguage_ArgChecks()
```
Comment (duncanp-sonar) [inline src/Integration.UnitTests/LanguageTests.cs]: These tests have been moved to a separate test class.

File: src/Integration.UnitTests/ProjectToLanguageMapperTests.cs [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+            vbNetProject.SetVBProjectKind();
+
+            // Act
+            var vbNetProjectLanguage = ProjectToLanguageMapper.GetLanguageForProject(vbNetProject);
+
+            // Assert
+            vbNetProjectLanguage.Should().Be(Language.VBNET, "Unexpected Language for VB project");
+        }
+
+        [Ignore] // not yet implemented
```
Comment (duncanp-sonar) [inline src/Integration.UnitTests/ProjectToLanguageMapperTests.cs]: The project type -> language mapper doesn't support .Net core projects yet so this test and the next one are currently disabled.

File: src/Integration/Language.cs [hunk 3]
```diff
@@ -38,13 +38,13 @@ namespace SonarLint.VisualStudio.Integration
...
     /// </remarks>
-    [DebuggerDisplay("{Name} (ID: {Id}, ProjectType: {ProjectType}, IsSupported: {IsSupported})")]
+    [DebuggerDisplay("{Name} (ID: {Id}, IsSupported: {IsSupported})")]
     [TypeConverter(typeof(LanguageConverter))]
     public sealed class Language : IEquatable<Language>
     {
         public readonly static Language Unknown = new Language();
-        public readonly static Language CSharp = new Language("CSharp", Strings.CSharpLanguageName, ProjectSystemHelper.CSharpProjectKind);
-        public readonly static Language VBNET = new Language("VB", Strings.VBNetLanguageName, ProjectSystemHelper.VbProjectKind);
+        public readonly static Language CSharp = new Language("CSharp", Strings.CSharpLanguageName);
```
Comment (duncanp-sonar) [inline src/Integration/Language.cs]: The _Language_ class used to have a one-one mapping from language -> project type. I've removed this.

File: src/Integration/LocalServices/ProjectSystemHelper.cs [hunk 4]
```diff
@@ -422,21 +426,6 @@ private static IEnumerable<IVsHierarchy> EnumerateProjects(IVsSolution solution)
             }
         }
 
-        public static bool IsVBProject(Project project)
```
Comment (duncanp-sonar) [inline src/Integration/LocalServices/ProjectSystemHelper.cs]: Dead code - wasn't being called from anywhere.

File: src/Integration/ProjectToLanguageMapper.cs [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    public static class ProjectToLanguageMapper
+    {
+        internal static readonly IDictionary<Guid, Language> KnownProjectTypes;
+
+        static ProjectToLanguageMapper()
+        {
+            KnownProjectTypes = new Dictionary<Guid, Language>();
+            KnownProjectTypes[new Guid(ProjectSystemHelper.CSharpProjectKind)] = Language.CSharp;
+            KnownProjectTypes[new Guid(ProjectSystemHelper.VbProjectKind)] = Language.VBNET;
+            // TODO: handle core project types
```
Comment (SonarTech) [inline src/Integration/ProjectToLanguageMapper.cs]: INFO Complete the task associated to this 'TODO' comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1135)

File: src/Integration/ProjectToLanguageMapper.cs [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+//           KnownProjectTypes[new Guid(ProjectSystemHelper.CSharpCoreProjectKind)] = Language.CSharp;
```
Comment (SonarTech) [inline src/Integration/ProjectToLanguageMapper.cs]: MAJOR Remove this commented out code. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS125)

File: src/Integration/Language.cs [hunk 7]
```diff
@@ -127,13 +115,8 @@ public static Language ForProject(EnvDTE.Project dteProject)
...
             }
 
-            Guid projectKind;
-            if (!Guid.TryParse(dteProject.Kind, out projectKind))
-            {
-                return Unknown;
-            }
-
-            return KnownLanguages.FirstOrDefault(x => x.ProjectType == projectKind) ?? Unknown;
+            // TODO: remove this method - change the callers to call the following method directly
```
Comment (SonarTech) [inline src/Integration/Language.cs]: INFO Complete the task associated to this 'TODO' comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1135)

File: src/Integration/LocalServices/ProjectSystemHelper.cs [hunk 8]
```diff
@@ -376,6 +378,8 @@ public IEnumerable<Guid> GetAggregateProjectKinds(IVsHierarchy hierarchy)
                 throw new ArgumentNullException(nameof(hierarchy));
             }
 
+            // TODO: is this relevant for core projects?
```
Comment (SonarTech) [inline src/Integration/LocalServices/ProjectSystemHelper.cs]: INFO Complete the task associated to this 'TODO' comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1135)

File: src/Integration/LocalServices/IProjectSystemHelperExtensions.cs [hunk 9]
```diff
@@ -31,6 +31,7 @@ internal static class IProjectSystemHelperExtensions
         /// </summary>
         public static bool IsKnownTestProject(this IProjectSystemHelper projectSystem, IVsHierarchy vsProject)
         {
+            //TODO: is this relevant for core projects?
```
Comment (SonarTech) [inline src/Integration/LocalServices/IProjectSystemHelperExtensions.cs]: INFO Complete the task associated to this 'TODO' comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1135)

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues
* ![MAJOR][MAJOR] 1 major
* ![INFO][INFO] 4 info

Watch the comments in this conversation to review them.

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: src/Integration/Language.cs [hunk 10]
```diff
@@ -127,13 +115,8 @@ public static Language ForProject(EnvDTE.Project dteProject)
...
+            return ProjectToLanguageMapper.GetLanguageForProject(dteProject);
```
Comment (duncanp-sonar) [inline src/Integration/Language.cs]: The method _Language.ForProject_ is called from many places in the code and tests. I haven't updated all of the callers yet to make these changes easier to review. I'll remove this method in a separate commit.

File: src/Integration/ProjectToLanguageMapper.cs [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+ */
+
+using System;
+using System.Collections.Generic;
+
+namespace SonarLint.VisualStudio.Integration
+{
+    public static class ProjectToLanguageMapper
+    {
+        internal static readonly IDictionary<Guid, Language> KnownProjectTypes;
```
Comment (Evangelink) [inline src/Integration/ProjectToLanguageMapper.cs]: It is usually better to make static initialization inline (compared to init in ctor).

File: src/Integration.UnitTests/ProjectToLanguageMapperTests.cs (hunk 2 above)
Comment (michalb-sonar) [reply src/Integration.UnitTests/ProjectToLanguageMapperTests.cs]: Please put this useful comment in the code.

File: src/Integration.UnitTests/ProjectToLanguageMapperTests.cs [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+            csProject.ProjectKind = ProjectSystemHelper.CSharpCoreProjectKind;
+
+            // Act
+            var csProjectLanguage = ProjectToLanguageMapper.GetLanguageForProject(csProject);
+
+            // Assert
+            csProjectLanguage.Should().Be(Language.CSharp, "Unexpected Language for C# Core project");
+        }
+
+        [Ignore] // not yet implemented
```
Comment (michalb-sonar) [inline src/Integration.UnitTests/ProjectToLanguageMapperTests.cs]: Also here, justification would be useful when we come back to it after n months.

File: src/Integration/ProjectToLanguageMapper.cs (hunk 5 above)
Comment (michalb-sonar) [reply src/Integration/ProjectToLanguageMapper.cs]: Why is this commented out?

File: src/Integration/ProjectToLanguageMapper.cs [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+
+namespace SonarLint.VisualStudio.Integration
+{
+    public static class ProjectToLanguageMapper
+    {
+        internal static readonly IDictionary<Guid, Language> KnownProjectTypes = new Dictionary<Guid, Language>()
+        {
+            { new Guid(ProjectSystemHelper.CSharpProjectKind), Language.CSharp },
+            { new Guid(ProjectSystemHelper.VbProjectKind), Language.VBNET },
+            // TODO: handle core project types
```
Comment (SonarTech) [inline src/Integration/ProjectToLanguageMapper.cs]: INFO Complete the task associated to this 'TODO' comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1135)

File: src/Integration/ProjectToLanguageMapper.cs [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+            //{ new Guid(ProjectSystemHelper.CSharpCoreProjectKind),  Language.CSharp },
+            //{ new Guid(ProjectSystemHelper.VbCoreProjectKind), Language.VBNET }
```
Comment (SonarTech) [inline src/Integration/ProjectToLanguageMapper.cs]: MAJOR Remove this commented out code. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS125)

File: src/Integration.UnitTests/ProjectToLanguageMapperTests.cs (hunk 12 above)
Comment (duncanp-sonar) [reply src/Integration.UnitTests/ProjectToLanguageMapperTests.cs]: @michalb-sonar "not yet implemented" is the justification, and there's a TODO in the product code. What comment would you prefer to see, and where?

Comment (michalb-sonar) [reply src/Integration.UnitTests/ProjectToLanguageMapperTests.cs]: Ah, ok. I assumed your [previos comment](https://github.com/SonarSource/sonarlint-visualstudio/pull/431/files#r150467145) means there is a 3rd party limitation - hence I said it's useful to say that. Guess that is not the case.

**Pull request merged**